src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs

```diff
@@ -71,13 +76,25 @@ public static SymbolKeyResolution Resolve(SymbolKeyReader reader, out string? fa
             private static void Resolve(
                 PooledArrayBuilder<INamedTypeSymbol> result,
                 INamespaceOrTypeSymbol container,
-                string metadataName,
+                string name,
                 int arity,
+                string? filePath,
                 bool isUnboundGenericType,
                 ITypeSymbol[] typeArguments)
             {
-                foreach (var type in container.GetTypeMembers(GetName(metadataName), arity))
+                foreach (var type in container.GetTypeMembers(name, arity))
                 {
+                    // if this is a 'file' type, then only resolve to a file-type from this same file
+                    if (filePath != null)
+                    {
+                        if (!type.IsFile ||
```

Comment:
Could there be a use case for returning a matching symbol if the file matches, but its not a file type anymore?  For example you delete the file keyword from the class, maybe in cases like FAR it would be useful to give it back the 'same' non-file symbol?